src/base/init/init.c

```diff
@@ -428,13 +428,13 @@ void low_mem_init(void)
   if (config.dpmi) {
     register_hardware_ram_virtual('U', DOSADDR_REL(ptr2), phys_rsv,
 	    LOWMEM_SIZE + HMASIZE);
-    /* create ext_mem alias for dpmi */
-    result = alias_mapping(MAPPING_EXTMEM, DOSADDR_REL(ptr2),
+  }
+  /* create ext_mem alias for int15/dpmi */
+  result = alias_mapping_pa(MAPPING_EXTMEM, LOWMEM_SIZE + HMASIZE,
```

Comment:
And suddenly its out of config.dpmi
block?

Reply:
yes, this is to associate SHM (`lowmem + LOWMEM_SIZE + HMASIZE`) to int15 memory for `$_dpmi=(0)`.

Without this PR, if `$_dpmi=(0)`, int15 memory is mapped by `physaddr_to_unixaddr` to `&mem_base[0x110000]`,
and else it's mapped to `&lowmem[0x110000]`. With this PR it's always mapped to SHM so it's more consistent.

Should I simply adjust the git commit log entry to mention that, or split the commit?

Reply:
Separate commit is always better.

Reply:
so I already had the seperate commit (number one in this series).

Just the comment is off. Actually I was confused and I need to change the comment again.

Basically I thought for $_dpmi != (0), int15 was copying to and from SHM (just like xms.c does), whereas in reality it copies in and out of mem_base like you wrote. The purpose of that alias_mapping was to connect that mem_base memory to SHM; with $_dpmi=(0) that SHM memory was allocated (together with lowmem) but not
used at all.

So in this PR I'm changing int15 to copy to and from SHM, but that change is independent of the setting of $_dpmi.

Then alias_mapping_pa sets up a) the connection between phys and SHM (via hwram_update_aliasmap) and b) creates the alias for DPMI if $_dpmi is not 0.

Reply:
> creates the alias for DPMI if $_dpmi is not 0.

And yet you move it out of the
config.dpmi check... So where is
the check so that it
"creates the alias for DPMI if $_dpmi is not 0"?

Reply:
I still don't understand.
Currently alias_mapping_pa() does this:
`addr2 = mappingdriver->alias(cap, MEM_BASE32(va), mapsize, protect, source);`
If you added some check in the
later patch, then please move it
into this patch. I don't understand
how `virtual mem_base addresses >= 0x110000 are
no longer accessed anywhere` if it
is accessed right in alias_mapping_pa().

Reply:
OK I think this all will stop confusing me
if you change the line 426 to use _virtual2(),
so that int15 is referred to lowmem_base,
in the patch before this one.

Reply:
So why not to use virtual2 at line
426 and leave that alias_mapping_pa()
as is, i.e. for dpmi only?

Reply:
Also reword doesn't help at all:
the patch is still called
`extmem: use alias_mapping_pa to map extmem`
and instead just moves the mapping
out of config.dpmi check.
I am about to ask you to split that
into 2 patches then, if you can't write
a log for this one. :)

Reply:
it does change from `alias_mapping` to `alias_mapping_pa` as well, it's not just moving it out of the dpmi check, and `alias_mapping_pa` (unlike `alias_mapping()`) calls `hwram_update_aliasmap()`, which is what is needed for extmem with or without dpmi.

so about the virtual2: if I use:
```
  register_hardware_ram_virtual2('X', LOWMEM_SIZE + HMASIZE, phys_rsv,
            ptr2, DOSADDR_REL(ptr2));
```
then this won't change anything, and the move to kvm_base two patches later won't move it.

On the other hand if I use
```
  register_hardware_ram_virtual2('X', LOWMEM_SIZE + HMASIZE, phys_rsv,
            lowmem + LOWMEM_SIZE + HMASIZE, DOSADDR_REL(ptr2));
```
then this is wrong since it includes XMS, not just extmem (the same issue we had before).

So I'm not quite sure what you mean with virtual2. Maybe it's better to split then?
```
  register_hardware_ram_virtual2('X', LOWMEM_SIZE + HMASIZE, EXTMEM_SIZE - HMASIZE,
            lowmem + LOWMEM_SIZE + HMASIZE, DOSADDR_REL(ptr2));
  register_hardware_ram_virtual('x', LOWMEM_SIZE + EXTMEM_SIZE, phys_rsv - (EXTMEM_SIZE - HMASIZE),
            DOSADDR_REL(ptr2));
```
However if I do that, once kvm_base is extended, the extmem part of it isn't used, and there will be fragmentation in the KVM slots, ie.
```
0 -> &kvm_base[0]
0x110000 -> LOWMEM(0x110000) (instead of 0x110000 -> &kvm_base[0x110000] -> LOWMEM(0x110000))
xms_base -> kvm_base[xms_base]
```
so then I need to undo the split again in the 3rd patch?

Reply:
> it does change from alias_mapping to alias_mapping_pa as well

Yes, and that part is properly described
in a log, so no questions to that one.

Yes, virtual2 stuff is not needed here, but
I don't understand why we have that
unchangelogged move, so please move
it to another patch then. :)

Reply:
it's now split into two patches. Unless you want me to squash patch 2 and patch 4 together?

Reply:
No, don't squash, I am trying to understand
this part.
So you say `but no longer accessed by int15()`
which probably applies to previous patch, not
this one. So for this patch I only see:
`This way the extmem mappings are fully consistent with XMS`.
Why not to write what does this mean?
I think that means "extmem now refers the
unix address (used by int15) to lowmem_base
also when dpmi disabled, while the previous
patch made it so only for dpmi enabled".
I surmise that from the prev patch, which has
now the updated description that makes things
clearer.

Reply:
I asked myself the same question and it's because update_aliasmap has this:
```
  if (dosaddr >= ALIAS_SIZE)
    return;

```

Reply:
So is this a regression in your prev
patch where you didn't change this
ALIAS_SIZE to eg mem_bases[MEM_BASE].size?
There is definitely some bug here.
Its not possible that you "refuse" to
document some changes or I can't
understand a single-line change for
so long, and no bugs around. :)

Reply:
no, it's not a regression. it has to be ALIAS_SIZE there.

in order to use `hwram_update_aliasmap`, `update_aliasmap` would need the physical address from the virtual one (i.e. the reverse of what we do now), and that has never been implemented.

Reply:
checking history:
859086eaa had this ok, obviously: `return LOWMEM(addr)`, and 4b970fc20d as well.
d54e82e3cd seems to have changed it.

Reply:
> and that has never been implemented.

Right, but seems rather simple:
[a.diff.gz](https://github.com/dosemu2/dosemu2/files/11020840/a.diff.gz)
What do you think?

Reply:
Forgot to check addr2 for -1 in the
above patch.

Reply:
yes that looks like a good idea! Can you do that as a seperate PR, and then I'll rebase on top of that?